doc/neps/nep-0057-array-in-array-out.rst

```diff
@@ -0,0 +1,231 @@
+downstream library test suites has been minimal in testing.
+
+Backward compatibility
+----------------------
+
+The motivation of this proposal is to eliminate the surprises associated with
+a scalar result when a zero-rank array would be expected. However, existing
+code may rely on the current behavior, and this presents backward compatibility
+concerns.
+
+The main concern for user code is that the mutable zero-rank arrays that
+replace immutable scalars are no longer hashable. For instance, they cannot
+by used directly as keys of dictionaries, the argument of an ``lru_cache``
+-decorated function, etc. In all circumstances, tbe patch is simple: convert
+the zero-rank array to a scalar with empty-tuple indexing.
+
+Running the test suites of dependent libraries against a branch of NumPy that
+implements these changes has revealed a few other issues. <Library maintainers,
+please summaries these issues here.>
+
+Detailed description
+--------------------
+
+The new functionality will be used in much the same was as old functionality,
+except that sometimes authors will need to convert zero-rank arrays to scalars
+rather than converting scalars to zero-rank arrays. For instance, consider the
+following:
+
+.. code:: python
+
+   import numpy as np
+   rng = np.random.default_rng(85878653462722874072976519960992129768)
+   x = rng.standard_normal(size=10)
+   y = np.sum(x, axis=-1)
+   z = {y: 'a duck'}  # use scalar result as dictionary key
+   y = np.asarray(y)  # convert to array to allow mutation
+   y[y < 0] = np.nan
+
+The use of ``x`` as a dictionary key would need to become ``x[()]``, but ``z``
+no longer needs to be explicitly converted to an array.
+
+.. code:: python
+
+   z = {y[()]: 'a duck'}  # extract scalar to use as dictionary key
+   y[y < 0] = np.nan      # no conversion to array required
+
+Realistic examples can be found throughout the codebases of dependent
+libraries.
+
+Related work
+------------
+
+All known libraries that attempt to implement the array API standard
+(e.g. ``cupy``, ``torch``, ``jax.numpy``, ``dask.array``) return
+zero-rank arrays as specified by the standard.
+
+Implementation
+--------------
+
+To implement the NEP, the branch prepared by Sebastian needs to be merged,
+and dependent libraries will need to prepare releases that adapt to (and take
+advantage of) the changes. Branches of libraries including SciPy and Matplotlib
+have already been prepared without much difficulty. Initially, users will have
+the option of opting into this behavior using an environment variable, so these
+releases need to be compatible with both old and new behaviors. To make the new
+behavior the default and only behavior, NumPy will need to advise users of the
+pending change and give the appropriate notice. The initial draft of this
+document does not specify the appropriate timeline and procedures; this line
+will be updated to reflect the consensus of the maintainer team.
```

Comment:
@seberg should I propose something? What would you suggest?

Reply:
Propose something.  I would say:
1. Create a branch and use it to adapt _and test_ downstream.
2. The final NEP acceptance will happen after this with potentially more discussion.

Or in other words, I would just not say much about the precise path of achieving step 2.  That makes sense IMO, because without 1 it is even harder to judge how big of a change it is (not that 1 will be a silver bullet).

How exactly you want to write it, I don't care.  You could even add a `.. note::` early on that a final decision is deferred until after further testing.